### PR TITLE
[InstCombine] Fix poison propagation in select of bitwise fold

### LINKED ISSUE #89500: [InstCombine] Miscompilation with poison vectors in `src_and_eq_C_xor_OrAndNotC`
Alive2: https://alive2.llvm.org/ce/z/yB3oAx
```ll
define <2 x i32> @src_and_eq_C_xor_OrAndNotC(<2 x i32> %0, <2 x i32> %1, <2 x i32> %2) {
entry:
  %and = and <2 x i32> %1, %0
  %cmp = icmp eq <2 x i32> %and, %2
  %xor = xor <2 x i32> %1, %0
  %or = or <2 x i32> %1, %0
  %not = xor <2 x i32> %2, <i32 -1, i32 poison>
  %and1 = and <2 x i32> %or, %not
  %cond = select <2 x i1> %cmp, <2 x i32> %xor, <2 x i32> %and1
  ret <2 x i32> %cond
}
```
```ll

----------------------------------------
define <2 x i32> @src_and_eq_C_xor_OrAndNotC(<2 x i32> %#0, <2 x i32> %#1, <2 x i32> %#2) {
entry:
  %and = and <2 x i32> %#1, %#0
  %cmp = icmp eq <2 x i32> %and, %#2
  %xor = xor <2 x i32> %#1, %#0
  %or = or <2 x i32> %#1, %#0
  %not = xor <2 x i32> %#2, { 4294967295, poison }
  %and1 = and <2 x i32> %or, %not
  %cond = select <2 x i1> %cmp, <2 x i32> %xor, <2 x i32> %and1
  ret <2 x i32> %cond
}
=>
define <2 x i32> @src_and_eq_C_xor_OrAndNotC(<2 x i32> %#0, <2 x i32> %#1, <2 x i32> %#2) nofree willreturn memory(none) {
entry:
  %or = or <2 x i32> %#1, %#0
  %not = xor <2 x i32> %#2, { 4294967295, poison }
  %and1 = and <2 x i32> %or, %not
  ret <2 x i32> %and1
}
Transformation doesn't verify!

ERROR: Target is more poisonous than source

Example:
<2 x i32> %#0 = < poison, #x00000000 (0) >
<2 x i32> %#1 = < poison, #x00000000 (0) >
<2 x i32> %#2 = < poison, #x00000000 (0) >

Source:
<2 x i32> %and = < poison, #x00000000 (0) >
<2 x i1> %cmp = < poison, #x1 (1) >
<2 x i32> %xor = < poison, #x00000000 (0) >
<2 x i32> %or = < poison, #x00000000 (0) >
<2 x i32> %not = < poison, poison >
<2 x i32> %and1 = < poison, poison >
<2 x i32> %cond = < poison, #x00000000 (0) >

Target:
<2 x i32> %or = < poison, #x00000000 (0) >
<2 x i32> %not = < poison, poison >
<2 x i32> %and1 = < poison, poison >
Source value: < poison, #x00000000 (0) >
Target value: < poison, poison >

Summary:
  0 correct transformations
  1 incorrect transformations
  0 failed-to-prove transformations
  0 Alive2 errors
```

Found with https://github.com/dtcxzyw/llvm-tools.
cc @nikic

### PR DESCRIPTION
**Pull request by nikic**
We're replacing the select with the false value here, but it may be more poisonous if m_Not contains poison elements. Fix this by introducing a m_NotForbidPoison matcher and using it here.

Fixes https://github.com/llvm/llvm-project/issues/89500.

### CODE REVIEW

File: llvm/include/llvm/IR/PatternMatch.h [hunk 1]
```diff
@@ -2596,6 +2602,13 @@ m_Not(const ValTy &V) {
   return m_c_Xor(m_AllOnes(), V);
 }
 
+template <typename ValTy>
+inline BinaryOp_match<cst_pred_ty<is_all_ones, false>, ValTy, Instruction::Xor,
+                      true>
+m_NotForbidPoison(const ValTy &V) {
+  return m_c_Xor(m_AllOnesForbidPoison(), V);
+}
```
Comment (goldsteinn) [inline llvm/include/llvm/IR/PatternMatch.h:2610]: missing unit tests?

Comment (nikic) [reply llvm/include/llvm/IR/PatternMatch.h:2610]: I prefer not to have unit tests for anything that has lit test coverage.

Comment (nikic) [reply llvm/include/llvm/IR/PatternMatch.h:2610]: (Except for exhaustive unit tests.)

Comment (goldsteinn) [reply llvm/include/llvm/IR/PatternMatch.h:2610]: Err I was confusing, not the `m_NotForbidPoison`, but for the `AllowPoison` template arg in the `cstval_pred_ty` unittests.

Either way though.

**Comment by goldsteinn** [comment]
LGTM w/ preference for unit tests for `AllowPoison`

**Review by dtcxzyw — APPROVED** [review]
LGTM.

**Pull request merged**